### ISSUE TITLE
False positive for `use_self` when using an associated type instead of the type itself.

### ISSUE THREAD

**Issue by DDOtten** [rust-lang/rust-clippy#2843]
A simple example of this is:

```rust
trait Foo {
    type Bar;
}

impl Foo for usize {
    type Bar = u8;
}

impl<T: Foo> Foo for Option<T> {
    type Bar = Option<T::Bar>;
}
```

[here is the playground link](https://play.rust-lang.org/?gist=733c15b6881eb762c0cd4c50b6970ebe&version=undefined&mode=undefined).

**Comment by flip1995** [comment]
Related to #1993. There is a WIP PR #1997 with a fix, but hasn't been completed for a few months.

**Comment by gnzlbg** [comment]
Another example:

```rust
#![warn(clippy::use_self)]

pub trait NonZeroT {
    type T;
}
impl NonZeroT for i32 {
    type T = std::num::NonZeroU32;
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=9926807d25af6d50d9b34139bd1c27ff
